src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/DictionaryLoggerScope.cs

```diff
@@ -37,54 +37,62 @@ public static DictionaryLoggerScope Current
 
         public static IDisposable Push(object state)
         {
-            IDictionary<string, object> stateValues;
-
-            if (state is IEnumerable<KeyValuePair<string, object>> stateEnum)
+            if (state is IDictionary<string, object> currentState)
+            {
+                BuildCurrentScope(currentState);
+            }
+            else if (state is IEnumerable<KeyValuePair<string, object>> stateEnum)
             {
-                // Convert this to a dictionary as we have scenarios where we cannot have duplicates. In this
-                // case, if there are dupes, the later entry wins.
+                IDictionary<string, object> stateValues;
+                // Convert this to a dictionary as we have scenarios where we cannot have duplicates.
+                // In this case, if there are dupes, the later entry wins.
                 stateValues = new Dictionary<string, object>();
```

Comment:
If you want even more performance out of this scope implementation, I recommend looking at [LoggerFactoryScopeProvider](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs) and basically following the `IExternalScopeProvider` pattern. It does a great job at avoiding extra allocations. For example here you are allocating an extra dictionary. And then `BuildCurrentScope` is going to allocate _another_ dictionary yet. `IExternalScopeProvider` avoids these allocations by not collapsing/processing the scope upfront, but rather letting consumers iterate through the scope values and they must perform type checking and consuming each scope level appropriately - and they also need to decide how to handle duplicates.

Not a blocker for this PR, but I thought I would mention this as the primary focus of this is performance.

Reply:
The dictionary allocation in the `Push` method is only if the state is `IEnumerable`. If it's already a dictionary, then there is no allocation. In all our code path we use a dictionary for state. `IEnumerable` is just to make sure we don't break if any customer is using it for scope.